worldedit-core/src/main/java/com/fastasyncworldedit/core/extent/clipboard/io/schematic/MinecraftStructure.java

```diff
@@ -1,273 +1,291 @@
+            if (properties != null) {
+                for (final Map.Entry<String, LinTag<?>> propertyPair : properties.value().entrySet()) {
+                    final Property<Object> property = blockType.getProperty(propertyPair.getKey());
+                    if (property == null) {
+                        continue;
                     }
-                    clipboard.setBlock(x, y, z, state);
+                    final String value = LinTagType.stringTag().cast(propertyPair.getValue()).value();
+                    block = block.with(property, property.getValueFor(value));
 
                 }
-            } catch (Exception e) {
-                e.printStackTrace();
             }
+            palette[i] = block;
         }
-        // Entities
-        ListTag entities = (ListTag) tags.get("entities");
-        if (entities != null) {
-            List<CompoundTag> entityList = (List<CompoundTag>) (List<?>) entities.getValue();
-            for (CompoundTag entityEntry : entityList) {
-                Map<String, Tag<?, ?>> entityEntryMap = entityEntry.getValue();
-                ListTag posTag = (ListTag) entityEntryMap.get("pos");
-                CompoundTag nbtTag = (CompoundTag) entityEntryMap.get("nbt");
-                String id = nbtTag.getString("Id");
-                Location location = NBTConversions.toLocation(clipboard, posTag, nbtTag.getListTag("Rotation"));
-                if (!id.isEmpty()) {
-                    BaseEntity state = new BaseEntity(EntityTypes.get(id), nbtTag);
-                    clipboard.createEntity(location, state);
-                }
+
+        // Blocks
+        final List<LinCompoundTag> blocks = parent.getListTag("blocks", LinTagType.compoundTag()).value();
+        for (final LinCompoundTag block : blocks) {
+            int state = block.getTag("state", LinTagType.intTag()).valueAsInt();
+            if (state >= palette.length) {
+                throw new IOException("state index exceeds palette length");
+            }
+            List<LinIntTag> pos = block.getTag("pos", LinTagType.listTag()).asTypeChecked(LinTagType.intTag()).value();
+            if (size.size() != 3) {
+                throw new IOException("Expected 'pos' to contain 3 integers, but got " + size.size());
             }
+            LinCompoundTag nbt = block.findTag("nbt", LinTagType.compoundTag());
+            if (nbt == null) {
+                clipboard.setBlock(pos.get(0).valueAsInt(), pos.get(1).valueAsInt(), pos.get(2).valueAsInt(), palette[state]);
+                continue;
+            }
+            clipboard.setBlock(
+                    pos.get(0).valueAsInt(), pos.get(1).valueAsInt(), pos.get(2).valueAsInt(),
+                    palette[state].toBaseBlock(nbt)
+            );
+        }
+
+        // Entities
+        LinListTag<@org.jetbrains.annotations.NotNull LinCompoundTag> entities = parent.findListTag("entities", LinTagType.compoundTag());
```

Comment:
Can you switch to the javax Nonnull annotation here? That's already imported

Reply:
Sadly not really possible, as for the javax annotations, generic types are not set as an allowed target. The only reason I've annotated them, is to get rid of the warning, as lin enforces nonnull there